crates/sui-core/src/execution_cache/utils.rs

```diff
@@ -0,0 +1,272 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::cmp::Ordering;
+use std::collections::VecDeque;
+
+use sui_types::base_types::SequenceNumber;
+
+/// CachedVersionMap is a map from version to value, with the additional contraints:
+/// - The key (SequenceNumber) must be monotonically increasing for each insert. If
+///   a key is inserted that is less than the previous key, it results in an assertion
+///   failure.
+/// - Similarly, only the item with the least key can be removed. If an item is removed
+///   from the middle of the map, it is marked for removal by setting its corresponding
+///   `should_remove` flag to true. If the item with the least key is removed, it is removed
+///   immediately, and any consecutive entries that are marked as `should_remove` are also
+///   removed.
+/// - The intent of these constraints is to ensure that there are never gaps in the collection,
+///   so that membership in the map can be tested by comparing to both the highest and lowest
+///   (first and last) entries.
+#[derive(Debug)]
+pub struct CachedVersionMap<V> {
+    values: VecDeque<(SequenceNumber, V)>,
+    should_remove: VecDeque<bool>,
+}
+
+impl<V> Default for CachedVersionMap<V> {
+    fn default() -> Self {
+        Self {
+            values: VecDeque::new(),
+            should_remove: VecDeque::new(),
+        }
+    }
+}
+
+impl<V> CachedVersionMap<V> {
+    pub fn is_empty(&self) -> bool {
+        self.values.is_empty()
+    }
+
+    pub fn insert(&mut self, version: SequenceNumber, value: V) {
+        assert!(
+            self.values.is_empty() || self.values.back().unwrap().0 < version,
+            "version must be monotonically increasing"
+        );
+        self.values.push_back((version, value));
+        self.should_remove.push_back(false);
+    }
+
+    pub fn all_lt_or_eq_rev<'a>(
+        &'a self,
+        version: &'a SequenceNumber,
+    ) -> impl Iterator<Item = &'a (SequenceNumber, V)> {
+        self.values
+            .iter()
+            .rev()
+            .take_while(move |(v, _)| v <= version)
+    }
+
+    pub fn get(&self, version: &SequenceNumber) -> Option<&V> {
+        if self.values.is_empty() {
+            return None;
+        }
+
+        for (v, value) in self.values.iter().rev() {
+            match v.cmp(version) {
+                Ordering::Less => return None,
+                Ordering::Equal => return Some(value),
```

Comment:
Should this also check `should_remove` flag?

Reply:
No - the cache can never have gaps. However your comment makes me realize that I think there may be a problem with the way i'm doing this right now.

Right now, if we have:  ```
dirty: [3, 4, 5]
cache: [1, 2]
```

and we commit version 4, we end up with ```
dirty: [3, 4*, 5]
cache: [1, 2, 4]
```

then if we commit 3, we will have an assertion failure.

instead, when moving from dirty to cache, i need to set the remove flag, then remove all entries with `should_remove: true` starting from the least. (This also removes the requirement of `Clone` on values in this map)

Reply:
I think I was a little confused by the word `remove`, and thinking more about a general scenario:

```
Say cache: [1, 2, 3]
cache.remove(2);

...... and some other parts of the code ......
cache.get(2) -> 2
cache.remove(1)
cache.get(2) -> None
```

One suggestion is maybe consider renaming/commenting the method `remove` to something that can reflect "mark the version for removal and return the list of the versions at the beginning of the queue that all can be removed", and there it can return `Option<Vec<V>>`.

Reply:
yeah i think i will change it to `committed` to indicate that it is no longer dirty

Reply:
so i tried to repro this in the tests and realized that its not really worth doing for now. Out-of-order commits are a bit risky since they could potentially leave the db in an impossible state. This is not a fundamental problem, we can always restore consistency during crash recovery by replaying all transactions, but it is too much to tackle for now.

But the good news is that if we only allow in-order commits for now, we don't need the dirty flag at all, because objects will only be removed from the low-version end of the queue.